### ISSUE TITLE
[Device Support Request]  Aqara Z1 Pro ZNQBKG45LM

### ISSUE THREAD

**Issue by HarvsG** [zigpy/zha-device-handlers#3127]
### Problem description

This device doesn't seem to be supported

### Solution description

Support to include, on off, different button presses, dimming, coupled vs uncoupled mode, and ideally bindings. It also supports energy monitoring and an anti flicker mode.

### Screenshots/Video

<details><summary>Screenshots/Video</summary>

[Paste/upload your media here]

</details>

### Device signature

<details><summary>Device signature</summary>

```json
[Paste the device signature here]
```

</details>

### Diagnostic information

<details><summary>Diagnostic information</summary>

```json
[Paste the diagnostic information here]
```

</details>

### Logs

<details><summary>Logs</summary>

```python
[Paste the logs here]
```

</details>

### Custom quirk

<details><summary>Custom quirk</summary>

```python
[Paste your custom quirk here]
```

</details>

### Additional information
https://www.aqara.com/en/product/smart-wall-switch-z1-pro-sea/
https://www.zigbee2mqtt.io/devices/ZNQBKG45LM.html
https://github.com/Koenkk/zigbee-herdsman-converters/pull/7160
Has 1 to 4 gang varieties

**Comment by kuminda** [comment]
I look forward to see ZHA integration for this device as I want to use the slider function

**Comment by KwFung7** [comment]
Please assist with supporting ZHA integration for the Z1 Pro.
Actually I’d like to contribute to this community, but I’m unsure where to begin. Thank you so much!

**Comment by KwFung7** [comment]
Found more references online:
[Example configuration from zigbee2mqtt](https://github.com/Koenkk/zigbee2mqtt/issues/21687)
[Model Info](https://opendoc.aqara.cn/docs/SDK%E5%BC%80%E5%8F%91/Android%E5%BC%80%E5%8F%91%E6%8C%87%E5%8D%97/%E8%AE%BE%E5%A4%87%E6%8E%A7%E5%88%B6/%E8%AE%BE%E5%A4%87%E6%8E%A7%E5%88%B6SDK%E6%94%AF%E6%8C%81%E8%AE%BE%E5%A4%87%E6%B8%85%E5%8D%95.html)

**Comment by brofeelgood** [comment]
May I know if these switches are already recognised in ZHA and if basic functions like on/off are working?

**Comment by KwFung7** [comment]
> May I know if these switches are already recognised in ZHA and if basic functions like on/off are working?

Yes, the on/off function work for me, but the slider entity didn't show up and the power consumption is always zero

<img width="757" alt="Image" src="https://github.com/user-attachments/assets/878f2d8e-08d8-40b0-9712-7181277af5bc" />

**Comment by brofeelgood** [comment]
> > May I know if these switches are already recognised in ZHA and if basic functions like on/off are working?
> 
> Yes, the on/off function work for me, but the slider entity didn't show up and the power consumption is always zero

I just found a [Youtube review](https://www.youtube.com/watch?v=C0XdbgY33es&t=833s) and it more or less corresponds to what you've shared. He did mention, however, that the power readings only come on after a minute or so.

I hope someone picks this up and works out a solution for it. Without the slider, this is just a Z1 non-pro switch.

**Comment by KwFung7** [comment]
I have created custom ZHA quirks that can help detect slider events from the Aqara Z1 Pro switch. Users can set up automation based on the received ZHA events. I plan to create a pull request to merge this into the zha-device-handlers repository. For those who are eager to test it before the official release, please follow the steps below to set it up on your local device:  
   
1. Add the following files to your Home Assistant **/config/custom_zha_quirks** directory (create it if it does not exist):  
   - [aqara_z1_pro_single.py](https://gist.github.com/KwFung7/4eeb4190cdabebc340c15a183a1ea928#file-aqara_z1_pro_single-py)
   - [aqara_z1_pro_double.py](https://gist.github.com/KwFung7/4eeb4190cdabebc340c15a183a1ea928#file-aqara_z1_pro_double-py)  
   - [aqara_z1_pro_triple.py](https://gist.github.com/KwFung7/4eeb4190cdabebc340c15a183a1ea928#file-aqara_z1_pro_triple-py)
   - [aqara_z1_pro_quadruple.py](https://gist.github.com/KwFung7/4eeb4190cdabebc340c15a183a1ea928#file-aqara_z1_pro_quadruple-py)
   
2. (Optional) Enable logging for custom ZHA quirks in **/config/configuration.yaml**:  
   ```yaml  
   logger:  
     default: info  
     logs:  
       zigpy: debug  
       homeassistant.components.zha: debug  
       zhaquirks: debug  
   ```  
   
3. Restart Home Assistant to activate the custom quirks.  
   
4. Verify that the quirks are loaded on the target devices and check the logs in **/config/home-assistant.log**.  
<img width="600" alt="Image" src="https://github.com/user-attachments/assets/575eb05d-4c13-4e93-a336-698edea7b1f8" />
   
5. Test the Aqara Z1 Pro slider and attempt to capture the ZHA slider event in Developer Tools.
<img width="600" alt="Image" src="https://github.com/user-attachments/assets/d2bc1f4b-77b9-4c15-9494-165679fba313" />

**Comment by oxelwhy** [comment]
I'm stuck in step4, unable to see the Quirk. My model is lumi.switch.acn059. Any idea?

**Comment by KwFung7** [comment]
@oxelwhy Did you find any log or error in **/config/home-assistant.log** related to **AqaraZ1ProTripleRockerSwitch**

`2025-05-17 14:46:46.169 DEBUG (SyncWorker_4) [switch_z1_pro_triple] AqaraZ1ProTripleRockerSwitch quirk module is being loaded! ZHA Profile ID: 0x0104, Device Type: 0x0000`

<img width="968" alt="Image" src="https://github.com/user-attachments/assets/6f26aa20-539f-4a58-9f0d-88800a336d65" />

**Comment by oxelwhy** [comment]
Nope, did not see any error. All I found is this, 
2025-05-18 13:55:23.373 DEBUG (SyncWorker_3) [aqara_z1_pro_double] AqaraZ1ProDoubleRockerSwitch quirk module is being loaded! ZHA Profile ID: 0x0104, Device Type: 0x0000
2025-05-18 13:55:23.378 DEBUG (SyncWorker_3) [aqara_z1_pro_quadruple] AqaraZ1ProQuadrupleRockerSwitch quirk module is being loaded! ZHA Profile ID: 0x0104, Device Type: 0x0000
2025-05-18 13:55:23.383 DEBUG (SyncWorker_3) [aqara_z1_pro_single] AqaraZ1ProSingleRockerSwitch quirk module is being loaded! ZHA Profile ID: 0x0104, Device Type: 0x0000
2025-05-18 13:55:23.390 DEBUG (SyncWorker_3) [aqara_z1_pro_triple] AqaraZ1ProTripleRockerSwitch quirk module is being loaded! ZHA Profile ID: 0x0104, Device Type: 0x0000

**Comment by KwFung7** [comment]
@oxelwhy How about this debug log ? if it doesn't show up then it prove the custom quirk was not being loaded properly.
`2025-05-18 17:05:08.061 DEBUG (MainThread) [switch_z1_pro_triple] AqaraZ1ProTripleRockerSwitch device initialized with IEEE:`

Besides you could try to compare your z1 provided endpoint in "Manage Zigbee device" with the signatures in added python.
<img width="989" alt="Image" src="https://github.com/user-attachments/assets/aa59c9bd-c133-4ecf-99e1-64f98960437a" />

**My "Manage Zigbee Device" signatures:**
```
{
  "node_descriptor": {
    "logical_type": 2,
    "complex_descriptor_available": 0,
    "user_descriptor_available": 0,
    "reserved": 0,
    "aps_flags": 0,
    "frequency_band": 8,
    "mac_capability_flags": 128,
    "manufacturer_code": 4447,
    "maximum_buffer_size": 82,
    "maximum_incoming_transfer_size": 82,
    "server_mask": 11264,
    "maximum_outgoing_transfer_size": 82,
    "descriptor_capability_field": 0
  },
  "endpoints": {
    "1": {
      "profile_id": "0x0104",
      "device_type": "0x0000",
      "input_clusters": [
        "0x0000",
        "0x0003",
        "0x0004",
        "0x0005",
        "0x0006",
        "0x0012",
        "0x0702",
        "0x0b04",
        "0xfcc0"
      ],
      "output_clusters": [
        "0x000a",
        "0x0019"
      ]
    },
    "2": {
      "profile_id": "0x0104",
      "device_type": "0x0000",
      "input_clusters": [
        "0x0004",
        "0x0005",
        "0x0006",
        "0x0012",
        "0xfcc0"
      ],
      "output_clusters": []
    },
    "3": {
      "profile_id": "0x0104",
      "device_type": "0x0000",
      "input_clusters": [
        "0x0004",
        "0x0005",
        "0x0006",
        "0x0012",
        "0xfcc0"
      ],
      "output_clusters": []
    },
    "4": {
      "profile_id": "0x0104",
      "device_type": "0x0000",
      "input_clusters": [
        "0xfcc0"
      ],
      "output_clusters": []
    },
    "21": {
      "profile_id": "0x0104",
      "device_type": "0x0000",
      "input_clusters": [
        "0x000c"
      ],
      "output_clusters": []
    }
  },
  "manufacturer": "Aqara",
  "model": "lumi.switch.acn058",
  "class": "switch_z1_pro_triple.AqaraZ1ProTripleRockerSwitch"
}
```

**Comment by KwFung7** [comment]
@oxelwhy I owned the single, double and triple rocker version but I don't have any quadruple version (acn059), that's why I just simply guessed the signature endpoint with the pattern I observed in other version. Perhaps there is something special in the quadruple version.

Could you send me the available signatures in your quadruple version so that I could help to cross check with my [aqara_z1_pro_quadruple.py](https://gist.github.com/KwFung7/4eeb4190cdabebc340c15a183a1ea928#file-aqara_z1_pro_quadruple-py)

<img width="990" alt="Image" src="https://github.com/user-attachments/assets/f5dccc70-c8bd-43ec-8034-4ebe780c52e5" />

**Comment by oxelwhy** [comment]
Ok, thanks for the guide. I've managed to get it recognized now. 

Updated the section as following.
```
            4: {
                PROFILE_ID: zha.PROFILE_ID,
                DEVICE_TYPE: zha.DeviceType.ON_OFF_SWITCH,
                INPUT_CLUSTERS: [
                    MultistateInputCluster,
                    MANUFACTURER_SPECIFIC_CLUSTER_ID,
                ],
                OUTPUT_CLUSTERS: [],
            },

```

**Comment by KwFung7** [comment]
@oxelwhy Done, I just found the misalignment, kindly copy the [aqara_z1_pro_quadruple.py](https://gist.github.com/KwFung7/4eeb4190cdabebc340c15a183a1ea928#file-aqara_z1_pro_quadruple-py) again and then restart your Home Assistant.

_**Changes:**_
<img width="1016" alt="Image" src="https://github.com/user-attachments/assets/4d53cffe-258b-45d8-9c64-5f3f8c758038" />

**Comment by oxelwhy** [comment]
Ok, tested your updated copy as well. It works. Thx much.

Btw, just out of curiosity, will the Sensors works as well?

![Image](https://github.com/user-attachments/assets/e9bdba05-ccdd-4f72-a06a-566350ce115b)

**Comment by KwFung7** [comment]
It does not work for me at the beginning and I'm not sure how to make it work, actually I'm not that familiar with ZHA, simply followed existing quirk to develop the script for z1

**Comment by oxelwhy** [comment]
No worry, thanks for your contribution and able to use slider. Will see if the wireless switch now able to work, so far single click only, not sure if double click etc will able to work.

**Comment by oxelwhy** [comment]
1. On the slider, how do you configure the slide up and slide down for automation? I want to archieve I.e. to set shade like 50% (drag bottom to middle), 25% (drag quarter from bottom) etc. Seems need a complicated or advanced script.
2. On the wireless switch, I notice Zigbee2mqtt can support single, double and hold. Current only support single, are you able to port it ?

**Comment by 1sakss0n** [comment]
Is anyone working on the Z1 Pro ZHA integration? I bought a quadruple rocker and would like to assist in any way I can.

**Comment by HarvsG** [comment]
Anyone managed to enable or disable decoupled or detached mode?

**Comment by shijaza-netizen** [comment]
@KwFung7 - Thanks for sharing this. I am curious if this quirk got merged with zha-device-handlers?

**Comment by KwFung7** [comment]
@shijaza-netizen No, my MR raised years ago but no one review and merge it, perhaps its not mature enough
